core/middleware.py

```diff
@@ -0,0 +1,15 @@
+import logging
+
+logging.basicConfig(filename="user_activity.log", encoding="utf-8", level=logging.DEBUG, format="%(asctime)s [%(levelname)s] - %(message)s", datefmt="%Y-%m-%d %H:%M:%S")
```

Comment:
Great! But we want to make this more configurable.  Having a file like this is perfect for development, but we'll want it to write out to stderr or stdout for when it runs on a production environment.

What I'd suggest is to alter this slightly to allow for that environment-level configuration. Something like:

```
from decouple import config
LOGLEVEL = config("DEBUG", default=INFO)
logging.basicConfig(level=LOGLEVEL, format="%(asctime)s [%(levelname)s] - %(message)s", datefmt="%Y-%m-%d %H:%M:%S")
```

Then, you could have a LOG_DESTINATION variable that, if set, would add in the filename and file encoding to the config as a condition afterwards.  This way, you can chose where things go depending on your requirements.

Reply:
The new changes centralise the logging config into settings.py and allow the log handler to be change based on a LOG_DESTINATION variable set in `.env`

When we decide on more destinations for the logs, we'll add them as handlers in the central config module. Then the new destinations can be mapped to their corresponding env var keywords by adding them to the `choose_log_handler` dictionary.

Is this what you expect?
Do tell me what you think about it.

Reply:
I see, I think this is close, but not quite what I had in mind. The way this works in production requires logging to be output to the `stdout` by default. Something [like this](https://odwyer.software/blog/logging-to-standard-output-with-django) maybe Saying that, logging to a file would be useful too, but should be a configurable thing. I think this might be able to be done using some branching. LOGGING, after all, is a dict.  Therefore, you can have a branch that sets the file or stdout config depending on environment variables.  

We also want to have the log level configurable so we can get more (or less) information if we need it without changing the code.

Putting it all together, and adapting what you updated the code to look like, you could do something like this:

```
# Get the log level
LOG_LEVEL = config("LOG_LEVEL", default="INFO")
LOG_DESTINATION = config("LOG_DESTINATION", default="file_output")
choose_log_handler = {"file_output": "file"}

# Set the logging configuration that should always be present
LOGGING = {
   'version': 1,
   'disable_existing_loggers': False,
   'formatters': {
       'verbose': {
           'format': '%(asctime)s %(name)-12s %(levelname)-8s %(message)s',
       },
   },
   'handlers': {
       'console': {
           'level': LOG_LEVEL ,
           'class': 'logging.StreamHandler',
           'stream': sys.stdout,
           'formatter': 'verbose'
       },
   },
   'loggers': {
       '': {
           'handlers': ['console'],
           'level': LOG_LEVEL ,
           'propagate': True,
       },
   },
}

# Optionally add in a file config if we need it
if LOG_DESTINATION:
   LOGGING["handlers"]["file"] = {
            "class": "logging.FileHandler",
            "filename": "user_activity.log",
            "level": LOG_LEVEL ,
            "formatter": "verbose",
        }
    LOGGING["loggers"]["core"] = {
            "handlers": [choose_log_handler[LOG_DESTINATION]],
            "level": "INFO",
        },
```

I've not tested it exactly, but that would give us a setup that's more configurable, but is guaranteed to do what we need in production (the output to stdout) while making development easier (optional file output)

Reply:
I've implemented the changes proposed.
  - Added the LOG_LEVEL variable and put it as a replacement of all hardcoded log-level values.
  - Explicit date formatting remains in the formatter specifications. It is not in your proposed paragraph so it can be removed if you prefer that.
  - I've included "console_output" in `choose_log_handler` so that if the dev user doesn't know that console is the default and then goes sets it again as a value of LOG_DESTINATION, it'll still work.